### PR TITLE
Improve ClrArray, add multi-dimensional APIs

### LINKED ISSUE #488: Support for multi dimensional arrays
ClrMD should handle multi-dimensional arrays, such as `string[,,]`.  At minimum we should expose a property like `ClrArray.Rank` to tell you whether it's an szarray or not.  Multi-dim arrays are not a common feature, so I'm not sure how much effort I want to spend on them.  Especially if there's not a lot of dac support for it.

### PR DESCRIPTION
**Pull request by NextTurn**
Closes #488

Please review the API shapes before I start working on them.

### CODE REVIEW

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 1]
```diff
@@ -111,6 +111,22 @@ public ClrArray(ulong address, ClrType type)
         /// <returns>An <see cref="int"/> hash code for this instance.</returns>
         public override int GetHashCode() => Address.GetHashCode();
 
+        public int GetLowerBound(int dimension) => throw new NotImplementedException(nameof(this.GetLowerBound));
+
+        public int GetUpperBound(int dimension) => throw new NotImplementedException(nameof(this.GetUpperBound));
+
+        // -or- GetValue
+        public T ReadValue<T>(int index) where T : unmanaged => throw new NotImplementedException(nameof(this.ReadValue));
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs]: I'm not sure if `Get` or `Read` is more clear.  For now I think using `GetValue` is right to be consistent with `ClrObject`, but if we decide to change it to Read later we should update `GetField` and similar as well.

**Comment by leculver** [comment]
This surface area seems good to me.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 2]
```diff
@@ -127,6 +117,139 @@ public int GetLength(int dimension)
...
+            }
+
+            int valueByteOffset = dataByteOffset + valueOffset * Type.ComponentSize;
+            return Type.ClrObjectHelpers.DataReader.Read<T>(Address + (ulong)valueByteOffset);
+        }
+
+        public ClrObject GetObjectValue(int index)
+        {
+            ulong address = GetValue<UIntPtr>(index).ToUInt64();
+            if (!Type.IsObjectReference)
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs]: nit:  Maybe this should go above the read?

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs]: Same for GetObjectValue.  It seems like we should throw an InvalidOperationException if the type isn't an object reference unconditionally.  It's possible that GetValue might throw a different exception depending on the underlying IMemoryReader implementation, so we should throw the InvalidOperationException as a priority.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 3]
```diff
@@ -127,6 +117,139 @@ public int GetLength(int dimension)
...
+                throw new InvalidOperationException();
+
+            return Type.Heap.GetObject(address);
+        }
+
+        public ClrObject GetObjectValue(params int[] indices)
+        {
+            ulong address = GetValue<UIntPtr>(indices).ToUInt64();
+            if (!Type.IsObjectReference)
+                throw new InvalidOperationException();
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs]: `throw new InvalidOperationException();` maybe this should be `throw new InvalidOperationException($"{Type} does not contain object references.");` or something similarly more descriptive.

**Comment by NextTurn** [comment]
> Still a draft?

I have some additional changes.

**Comment by NextTurn** [comment]
Do we want to check `sizeof(T) == Type.ComponentSize`?

**Comment by leculver** [comment]
> Do we want to check `sizeof(T) == Type.ComponentSize`?

Yes, since we are checking against the length/bounds of the array.  It prevents a useful case where you want to read it all as a byte array or something, but we could (eventually) provide other mechanisms to get that.

**Pull request merged**